**1.Data Collection**

**1.1. Imports**

In [267]:
import pandas as pd
import numpy as np
import os


**1.2.Load Data**

In [268]:
asthma_adults= pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(adults)%20(full%20table).csv")
asthma_age_4= pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(age%204%20and%20under)%20(full%20table).csv")
asthma_age5_17=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Asthma%20emergency%20department%20visits%20(age%205%20to%2017)%20(full%20table).csv")
pm2_5=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Fine%20particles%20(PM%202.5)%20(full%20table).csv")
no2=pd.read_csv("https://raw.githubusercontent.com/Mary080102/Capston-Two/main/Raw%20data/NYC%20EH%20Data%20Portal%20-%20Nitrogen%20dioxide%20(NO2)%20(full%20table).csv")

**2.Data Definition**

In [269]:
#getting a quick overview of the DataFrame
asthma_adults.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792 entries, 0 to 791
Data columns (total 8 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        792 non-null    int64 
 1   GeoType                           792 non-null    object
 2   GeoID                             792 non-null    int64 
 3   GeoRank                           792 non-null    int64 
 4   Geography                         792 non-null    object
 5   Age-adjusted rate per 10,000      791 non-null    object
 6   Estimated annual rate per 10,000  792 non-null    object
 7   Number                            792 non-null    object
dtypes: int64(3), object(5)
memory usage: 49.6+ KB


In [302]:
asthma_adults.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

  Age-adjusted rate per 10,000 Estimated annual rate per 10,000 Number  
0                         17.9                             14.6     80  
1                          9.8                              9.7     81  
2                         44.9                             43.9    661  
3                         28.6                             26.6    364  
4                         21.5                             20.0    116

In [270]:
asthma_age_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839 entries, 0 to 838
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        839 non-null    int64 
 1   GeoType                           839 non-null    object
 2   GeoID                             839 non-null    int64 
 3   GeoRank                           839 non-null    int64 
 4   Geography                         839 non-null    object
 5   Estimated annual rate per 10,000  839 non-null    object
 6   Number                            839 non-null    object
dtypes: int64(3), object(4)
memory usage: 46.0+ KB


In [368]:
asthma_age_4.head()

        

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

  Estimated annual rate per 10,000 Number  
0                            16.6*     7*  
1                             3.2*     1*  
2                             73.0     47  
3                             23.6     11  
4                            32.5*     7*

In [271]:
asthma_age5_17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 839 entries, 0 to 838
Data columns (total 7 columns):
 #   Column                            Non-Null Count  Dtype 
---  ------                            --------------  ----- 
 0   TimePeriod                        839 non-null    int64 
 1   GeoType                           839 non-null    object
 2   GeoID                             839 non-null    int64 
 3   GeoRank                           839 non-null    int64 
 4   Geography                         839 non-null    object
 5   Estimated annual rate per 10,000  839 non-null    object
 6   Number                            839 non-null    object
dtypes: int64(3), object(4)
memory usage: 46.0+ KB


In [432]:
asthma_age5_17.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

  Estimated annual rate per 10,000 Number  
0                            11.9*     6*  
1                             8.8*     5*  
2                             81.9    136  
3                             42.1     34  
4                            19.4*     5*

In [272]:
pm2_5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4554 entries, 0 to 4553
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   TimePeriod              4554 non-null   object 
 1   GeoType                 4554 non-null   object 
 2   GeoID                   4554 non-null   int64  
 3   GeoRank                 4554 non-null   int64  
 4   Geography               4554 non-null   object 
 5   10th percentile mcg/m3  4554 non-null   float64
 6   90th percentile mcg/m3  4554 non-null   float64
 7   Mean mcg/m3             4554 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 284.8+ KB


In [451]:
pm2_5.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile mcg/m3  \
0             Financial District (CD1)                     6.7   
1     Greenwich Village and Soho (CD2)                     7.6   
2  Lower East Side and Chinatown (CD3)                     6.1   
3            Clinton and Chelsea (CD4)                     6.7   
4                        Midtown (CD5)                     8.2   

   90th percentile mcg/m3  Mean mcg/m3  
0                     7.8          7.2  
1                     9.3          8.4  
2                     8.6          7.2  
3                     9.0          7.8  
4                     9.7          9.1

In [273]:
no2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4554 entries, 0 to 4553
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   TimePeriod           4554 non-null   object 
 1   GeoType              4554 non-null   object 
 2   GeoID                4554 non-null   int64  
 3   GeoRank              4554 non-null   int64  
 4   Geography            4554 non-null   object 
 5   10th percentile ppb  4554 non-null   float64
 6   90th percentile ppb  4554 non-null   float64
 7   Mean ppb             4554 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 284.8+ KB


In [453]:
no2.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile ppb  \
0             Financial District (CD1)                 19.3   
1     Greenwich Village and Soho (CD2)                 19.2   
2  Lower East Side and Chinatown (CD3)                 15.8   
3            Clinton and Chelsea (CD4)                 18.6   
4                        Midtown (CD5)                 22.4   

   90th percentile ppb  Mean ppb  
0                 22.4      21.0  
1                 21.1      20.2  
2                 19.9      17.8  
3                 24.4      21.4  
4                 27.8      25.5

Based on the results of .info() method, data type of some of columns must be changede. For example TimePeriod in pm2_5 dataset and no2 dataset are object that should change to int. data type of Estimated annual rate per 10,000, and Age-adjusted rate per 10,000 are object that must be foat. Data type of Number is object that must be int. After run .astype() for change data type I got error the dataset contain charector '†'. So first of all I have to find how many of this character are there? and handle it.

**2.1.Change data type of asthma_adults dataset**

In [329]:
asthma_adults["Age-adjusted rate per 10,000"].str.contains('†').sum()

1

there is exactly one entry in the "Age-adjusted rate per 10,000" column of the asthma_adults DataFrame that contains the symbol '†'.

In [330]:
# Create a copy of the DataFrame
asthma_adults_1 = asthma_adults.copy()


In [331]:
asthma_adults_1["Age-adjusted rate per 10,000"].isna().sum()

1

In [332]:
# replace '†' with empty string, then convert to numeric, setting errors='coerce'
asthma_adults_1["Age-adjusted rate per 10,000"] = pd.to_numeric(asthma_adults_1["Age-adjusted rate per 10,000"].str.replace('†', ''), errors='coerce')


In [333]:
# Calculate the mean of the column, excluding NaN values
mean_value= round(asthma_adults_1["Age-adjusted rate per 10,000"].mean(),2)

In [334]:
print(mean_value)

96.83


In [335]:
asthma_adults_1["Age-adjusted rate per 10,000"].isna().sum()

2

In [340]:
# Fill NaN values with the mean value
asthma_adults_1["Age-adjusted rate per 10,000"].fillna(mean_value, inplace=True)

In [341]:
asthma_adults_1["Age-adjusted rate per 10,000"].isna().sum()

0

In [342]:
#convert column "Estimated annual rate per 10,000" to numeric, setting errors='coerce'
asthma_adults_1["Estimated annual rate per 10,000"] = pd.to_numeric(asthma_adults_1["Estimated annual rate per 10,000"], errors='coerce')


In [343]:
asthma_adults_1["Estimated annual rate per 10,000"].isna().sum()

1

There is one entry with NaN value. I want to replace it with mean value.

In [344]:
# Calculate the mean of the column, excluding NaN values
mean_value_es=round(asthma_adults_1["Estimated annual rate per 10,000"].mean(),2)

In [345]:
print(mean_value_es)

95.58


In [346]:
# Fill NaN values with the mean value
asthma_adults_1["Estimated annual rate per 10,000"].fillna(mean_value_es, inplace=True)

In [349]:
# Using .value_counts()to see if there are any unusual values in column"Number". 
asthma_adults_1["Number"].value_counts()

520       4
494       3
982       3
1,348     3
740       3
         ..
21,089    1
20,608    1
12,872    1
9,598     1
62,425    1
Name: Number, Length: 691, dtype: int64

The result showed that some of entries has (,). I ahve to remove the comma and then change the type od data.

In [350]:
#Remove the ','  from the 'Number' in asthma_adults_1 dataset.
asthma_adults_1["Number"]= asthma_adults_1["Number"].str.replace(",", "")

In [351]:
#convert column "Number" to numeric, setting errors='coerce'
asthma_adults_1["Number"]= pd.to_numeric(asthma_adults_1["Number"], errors="coerce")

In [352]:
asthma_adults_1["Number"].isna().sum()

1

Result shows there is one entry with NaN value and it will be replace with mean value.

In [353]:
#Calculate the mean of the column "Number", excluding NaN values
mean_value_num= round(asthma_adults_1["Number"].mean(),2)

In [354]:
print(mean_value_num)

3608.03


In [355]:
# Fill NaN values with the mean value
asthma_adults_1["Number"].fillna(mean_value_num, inplace=True)

In [357]:
asthma_adults_1.sample(10)

TimePeriod   GeoType  GeoID  GeoRank  \
454        2012   Borough      5        1   
293        2016     UHF42    403        4   
44         2020        CD    403        6   
617        2008     UHF42    211        4   
304        2016     UHF42    504        4   
7          2020        CD    108        6   
213        2019  Citywide      1        0   
137        2019        CD    307        6   
373        2013     UHF42    207        4   
78         2020     UHF42    302        4   

                                Geography  Age-adjusted rate per 10,000  \
454                         Staten Island                          85.9   
293                  Flushing - Clearview                          24.6   
44                  Jackson Heights (CD3)                          18.4   
617               Williamsburg - Bushwick                         225.5   
304             South Beach - Tottenville                          29.0   
7                   Upper East Side (CD8)                          11.2   
213                         New York City                          78.3   
137                     Sunset Park (CD7)                          42.6   
373              East Flatbush - Flatbush                         139.9   
78   Central Harlem - Morningside Heights                         127.1   

     Estimated annual rate per 10,000   Number  
454                              84.3   3069.0  
293                              23.3    511.0  
44                               18.0    258.0  
617                             220.8   3282.0  
304                              27.0    415.0  
7                                10.1    198.0  
213                              77.1  54166.0  
137                              40.2    425.0  
373                             139.5   3274.0  
78                              126.3   1915.0

**2.2.Change data type of asthma_age_4 dataset**

Based on the results of .info(), columns of " Estimated annual rate per 10,000" and "Number" must be change to numeric data type.

In [386]:
# Create a copy of the DataFrame
asthma_age_4_1=asthma_age_4.copy()

In [390]:
asthma_age_4_1.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

  Estimated annual rate per 10,000 Number  
0                            16.6*     7*  
1                             3.2*     1*  
2                             73.0     47  
3                             23.6     11  
4                            32.5*     7*

In [391]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000". 
asthma_age_4_1["Estimated annual rate per 10,000"].value_counts()

23.6     3
99.1     3
72.3     2
188.8    2
155.5    2
        ..
316.4    1
183.2    1
198.7    1
284.7    1
377.2    1
Name: Estimated annual rate per 10,000, Length: 784, dtype: int64

In [392]:
# Due to the type of the column is object at first we need to convert the column to string and 
#them remove * from the number
asthma_age_4_1["Estimated annual rate per 10,000"]=asthma_age_4_1["Estimated annual rate per 10,000"].astype(str)

In [395]:
# Replace '*' with '' before converting to numeric
asthma_age_4_1["Estimated annual rate per 10,000"]=asthma_age_4_1["Estimated annual rate per 10,000"].str.replace('*', '',regex=True)

In [396]:
#convert column to numeric, setting errors='coerce'
asthma_age_4_1["Estimated annual rate per 10,000"]= pd.to_numeric(asthma_age_4_1["Estimated annual rate per 10,000"], errors='coerce')

In [397]:
asthma_age_4_1["Estimated annual rate per 10,000"].dtypes

dtype('float64')

In [398]:
asthma_age_4_1["Estimated annual rate per 10,000"].isna().sum()

9

Result shows that there 9 entry with NaN values in column "Estimated annual rate per 10,000". These have to be replace with mean value.

In [399]:
#Calculate the mean of the column "Estimated annual rate per 10,000", excluding NaN values
mean_age_4= round(asthma_age_4_1["Estimated annual rate per 10,000"].mean(),2)

In [400]:
print(mean_age_4)

268.19


In [401]:
# Fill NaN values with the mean value
asthma_age_4_1["Estimated annual rate per 10,000"].fillna(mean_age_4, inplace=True)

In [403]:
asthma_age_4_1["Estimated annual rate per 10,000"].isna().sum()

0

In [404]:
asthma_age_4_1["Estimated annual rate per 10,000"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 839 entries, 0 to 838
Series name: Estimated annual rate per 10,000
Non-Null Count  Dtype  
--------------  -----  
839 non-null    float64
dtypes: float64(1)
memory usage: 6.7 KB


Result shows Dtype of "Estimated annual rate per 10,000" column change to float.

In [408]:
# Using .value_counts()to see if there are any unusual values in column"Number".
asthma_age_4_1["Number"].value_counts().tail(10)

571       1
844       1
211       1
1,225     1
111       1
1,024     1
1,454     1
1,070     1
1,561     1
19,927    1
Name: Number, dtype: int64

It looks like the column has number with comma, so I have to remove from data entry and then change the data type.

In [409]:
#Remove the comma from the 'Number' column
asthma_age_4_1["Number"]= asthma_age_4_1["Number"].str.replace(",", "")

In [412]:
asthma_age_4_1["Number"].value_counts().sample(20)

759     1
361     1
126     1
396     2
500     2
296     3
6*      1
1717    1
212     2
47      3
621     1
350     1
1169    1
647     1
7939    1
309     1
167     1
210     1
806     1
343     2
Name: Number, dtype: int64

In [414]:
asthma_age_4_1["Number"].dtype

dtype('O')

Based on the results of sample there is * in the column. So we need to remove that, but the dtype of the column is objest. So we need to convert the column to string first and then remove * from data.

In [417]:
#Convert the column to string first
asthma_age_4_1["Number"]=asthma_age_4_1["Number"].astype(str)
#Remove asterisks
asthma_age_4_1["Number"]=asthma_age_4_1["Number"].str.replace('*','',regex=False)

In [419]:
#convert column to numeric, setting errors='coerce'
asthma_age_4_1["Number"]= pd.to_numeric(asthma_age_4_1["Number"], errors="coerce")

In [421]:
asthma_age_4_1["Number"].dtype

dtype('int64')

In [420]:
asthma_age_4_1["Number"].isna().sum()

0

There are no any null value.

In [425]:
asthma_age_4_1.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

   Estimated annual rate per 10,000  Number  
0                              16.6       7  
1                               3.2       1  
2                              73.0      47  
3                              23.6      11  
4                              32.5       7

In [426]:
asthma_age_4_1["Number"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 839 entries, 0 to 838
Series name: Number
Non-Null Count  Dtype
--------------  -----
839 non-null    int64
dtypes: int64(1)
memory usage: 6.7 KB


Result shows Dtype of "Number" has changed to integer.

**2.3.Change data type of asthma_age5_17 dataset**

 Columns "Estimated annual rate per 10,000" and "Number" must change to numeric data type.But before that we need to remove(,) and (*) form the data.

In [427]:
# Create a copy of the DataFrame
asthma_age5_17_1= asthma_age5_17.copy()

In [428]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000".
asthma_age5_17_1["Estimated annual rate per 10,000"].value_counts()

139.9    3
47.8     3
34.6     3
98.5     3
160.0    3
        ..
78.3     1
22.4     1
120.2    1
186.4    1
453.7    1
Name: Estimated annual rate per 10,000, Length: 737, dtype: int64

In [434]:
#Convert the column to string first
asthma_age5_17_1["Estimated annual rate per 10,000"]=asthma_age5_17_1["Estimated annual rate per 10,000"].astype(str)
#Remove asterisks
asthma_age5_17_1["Estimated annual rate per 10,000"]=asthma_age5_17_1["Estimated annual rate per 10,000"].str.replace('*','',regex=False)

In [435]:
#convert column to numeric, setting errors='coerce'
asthma_age5_17_1["Estimated annual rate per 10,000"]= pd.to_numeric(asthma_age5_17_1["Estimated annual rate per 10,000"], errors="coerce")


In [436]:
asthma_age5_17_1["Estimated annual rate per 10,000"].isna().sum()

0

There are no entries with NaN values.

In [437]:
# Using .value_counts()to see if there are any unusual values in column"Estimated annual rate per 10,000".
asthma_age5_17_1["Number"].value_counts().head(20)

109      6
113      5
5*       4
29       4
135      4
116      4
139      4
253      4
202      4
81       4
94       4
85       4
327      3
351      3
37       3
244      3
1,209    3
321      3
348      3
291      3
Name: Number, dtype: int64

Results shows that some number are with * and some of them have comma. (*) in dataset means that **Estimate is based on small numbers so should be interpreted with caution**. So I want to add new column to asthma_age5_17_1 that will indicate whether the original value had an asterisk. This way, I can maintain the information about the reliability of the estimate.

In [438]:
#Create a new column to indicate estimates based on small numbers
asthma_age5_17_1["small_number_estimate"]= asthma_age5_17_1["Number"].str.contains("\*")

In [439]:
asthma_age5_17_1.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0        2020      CD    101        6             Financial District (CD1)   
1        2020      CD    102        6     Greenwich Village and Soho (CD2)   
2        2020      CD    103        6  Lower East Side and Chinatown (CD3)   
3        2020      CD    104        6            Clinton and Chelsea (CD4)   
4        2020      CD    105        6                        Midtown (CD5)   

   Estimated annual rate per 10,000 Number  small_number_estimate  
0                              11.9     6*                   True  
1                               8.8     5*                   True  
2                              81.9    136                  False  
3                              42.1     34                  False  
4                              19.4     5*                   True

New column "small_number_estimate" added to the dataset and shows which number's estimate is based on small numbers so should be interpreted with caution.

In [441]:
#Convert the column to string first
asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].astype(str)

# Remove the asterisks and commas
asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].str.replace("*", "",regex=False)


In [443]:
asthma_age5_17_1["Number"]=asthma_age5_17_1["Number"].str.replace(",","")

In [444]:
asthma_age5_17_1["Number"].value_counts()

109      6
113      5
5        4
29       4
135      4
        ..
228      1
1052     1
444      1
16       1
25834    1
Name: Number, Length: 614, dtype: int64

In [445]:
#convert column to numeric, setting errors='coerce'
asthma_age5_17_1["Number"]= pd.to_numeric(asthma_age5_17_1["Number"], errors="coerce")

In [446]:
asthma_age5_17_1["Number"].isna().sum()

0

**2.4.Change data type of pm2_5 dataset**

The "TimePeriod" column has object data type that has to change to numeric.

In [454]:
# Creat a copy of dataFrame
pm2_5_1= pm2_5.copy()

In [455]:
# Using .value_counts()to see if there are any unusual values in column"TimePeriod".
pm2_5_1["TimePeriod"].value_counts()

Annual Average 2022    141
Summer 2018            141
Winter 2014-15         141
Summer 2015            141
Summer 2022            141
Winter 2015-16         141
Summer 2016            141
Annual Average 2016    141
Winter 2016-17         141
Summer 2017            141
Annual Average 2017    141
Winter 2017-18         141
Annual Average 2015    141
Annual Average 2018    141
Annual Average 2020    141
Winter 2018-19         141
Winter 2021-22         141
Summer 2021            141
Winter 2020-21         141
Annual Average 2021    141
Summer 2020            141
Winter 2019-20         141
Annual Average 2019    141
Summer 2019            141
Annual Average 2011     65
Summer 2009             65
Annual Average 2009     65
Winter 2009-10          65
Summer 2010             65
Annual Average 2010     65
Winter 2010-11          65
Summer 2011             65
Annual Average 2013     65
Winter 2011-12          65
Summer 2012             65
Annual Average 2012     65
Winter 2012-13          65
S

Results of .value_counts() shows different type of entry including: Annual Average, Summer, and Winter. Because values of the other dataframes sucha as asthma_adults, asthma_age_4, and asthma_5_17 are based on annual not seasonal, therefore I decide to keep just Annual Average values.

In [456]:
# filter rows to keep Annual Average rows
pm2_5_sub= pm2_5_1[pm2_5_1["TimePeriod"].str.startswith("Annual Average")]

In [457]:
pm2_5_sub.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile mcg/m3  \
0             Financial District (CD1)                     6.7   
1     Greenwich Village and Soho (CD2)                     7.6   
2  Lower East Side and Chinatown (CD3)                     6.1   
3            Clinton and Chelsea (CD4)                     6.7   
4                        Midtown (CD5)                     8.2   

   90th percentile mcg/m3  Mean mcg/m3  
0                     7.8          7.2  
1                     9.3          8.4  
2                     8.6          7.2  
3                     9.0          7.8  
4                     9.7          9.1

All data for "TimePeriod" column just include Annual Average.

In [458]:
#Remove the 'Annual' prefix from the 'TimePeriod' in pm2_5_sub
pm2_5_sub= pm2_5_sub.copy()
pm2_5_sub["TimePeriod"]= pm2_5_sub["TimePeriod"].str.replace("Annual Average", "")

In [459]:
pm2_5_sub.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0       2022      CD    101        6             Financial District (CD1)   
1       2022      CD    102        6     Greenwich Village and Soho (CD2)   
2       2022      CD    103        6  Lower East Side and Chinatown (CD3)   
3       2022      CD    104        6            Clinton and Chelsea (CD4)   
4       2022      CD    105        6                        Midtown (CD5)   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  
0                     6.7                     7.8          7.2  
1                     7.6                     9.3          8.4  
2                     6.1                     8.6          7.2  
3                     6.7                     9.0          7.8  
4                     8.2                     9.7          9.1

In [464]:
pm2_5_sub.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1518 entries, 0 to 4423
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   TimePeriod              1518 non-null   object 
 1   GeoType                 1518 non-null   object 
 2   GeoID                   1518 non-null   int64  
 3   GeoRank                 1518 non-null   int64  
 4   Geography               1518 non-null   object 
 5   10th percentile mcg/m3  1518 non-null   float64
 6   90th percentile mcg/m3  1518 non-null   float64
 7   Mean mcg/m3             1518 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 106.7+ KB


In [460]:
#Creat a copy of dataframe
no2_1= no2.copy()

In [191]:
# filter rows to keep Annual Average rows
no2_sub= no2_1[no2_1["TimePeriod"].str.startswith("Annual Average")]
no2_sub.head()

TimePeriod GeoType  GeoID  GeoRank  \
0  Annual Average 2022      CD    101        6   
1  Annual Average 2022      CD    102        6   
2  Annual Average 2022      CD    103        6   
3  Annual Average 2022      CD    104        6   
4  Annual Average 2022      CD    105        6   

                             Geography  10th percentile ppb  \
0             Financial District (CD1)                 19.3   
1     Greenwich Village and Soho (CD2)                 19.2   
2  Lower East Side and Chinatown (CD3)                 15.8   
3            Clinton and Chelsea (CD4)                 18.6   
4                        Midtown (CD5)                 22.4   

   90th percentile ppb  Mean ppb  
0                 22.4      21.0  
1                 21.1      20.2  
2                 19.9      17.8  
3                 24.4      21.4  
4                 27.8      25.5

In [192]:
#Remove the 'Annual' prefix from the 'TimePeriod' in pm2_5_sub
no2_sub=no2_sub.copy()
no2_sub["TimePeriod"]= no2_sub["TimePeriod"].str.replace("Annual Average", "")
no2_sub.head()

TimePeriod GeoType  GeoID  GeoRank                            Geography  \
0       2022      CD    101        6             Financial District (CD1)   
1       2022      CD    102        6     Greenwich Village and Soho (CD2)   
2       2022      CD    103        6  Lower East Side and Chinatown (CD3)   
3       2022      CD    104        6            Clinton and Chelsea (CD4)   
4       2022      CD    105        6                        Midtown (CD5)   

   10th percentile ppb  90th percentile ppb  Mean ppb  
0                 19.3                 22.4      21.0  
1                 19.2                 21.1      20.2  
2                 15.8                 19.9      17.8  
3                 18.6                 24.4      21.4  
4                 22.4                 27.8      25.5

In [465]:
# Change the data type into int
pm2_5_sub["TimePeriod"]=pm2_5_sub["TimePeriod"].astype(int)

no2_sub["TimePeriod"]=no2_sub["TimePeriod"].astype(int)

In [466]:

no2_sub["TimePeriod"].info()

<class 'pandas.core.series.Series'>
Int64Index: 1518 entries, 0 to 4423
Series name: TimePeriod
Non-Null Count  Dtype
--------------  -----
1518 non-null   int64
dtypes: int64(1)
memory usage: 23.7 KB


Data type on the "TimePeriod" column changed to int.

**3.Data Cleaning**

**3.1. Dropping Unneccessary Rows:**

I want to choose three neighborhoods in New York City to assess air quality that can provide an overview of the whole area requires considering diversity in geography, urban density, and potential sources of pollution. Based on the Geography column in csv files that includes all neighborhoods,I want to choose New yourk, Manhatan,Staten Island.

In [467]:
# List of names I want to keep
rows_to_keep=["New York City","Manhattan","Staten Island"]

In [468]:
# Creat a new Dataframe with only rows for our study with the name,"New York City","Manhattan","Staten Island"
asthma_adults_df=asthma_adults_1.loc[asthma_adults_1["Geography"].isin(rows_to_keep)]

asthma_age_4_df= asthma_age_4_1.loc[asthma_age_4_1["Geography"].isin(rows_to_keep)]

asthma_age5_17_df=asthma_age5_17_1.loc[asthma_age5_17_1["Geography"].isin(rows_to_keep)]

pm2_5_df= pm2_5_sub.loc[pm2_5_sub["Geography"].isin(rows_to_keep)]

no2_df= no2_sub.loc[no2_sub["Geography"].isin(rows_to_keep)]

**3.2.Dropping unneccessary Columns**

In [197]:
# finding common column to merge
common_columns= set(asthma_adults_df.columns)& set(asthma_age_4_df.columns)&set(asthma_age5_17_df.columns)&set(pm2_5_df.columns)&set(no2_df.columns)
print(common_columns)

{'Geography', 'GeoRank', 'GeoID', 'GeoType', 'TimePeriod'}


These are common columns in all five dataFrames. I can choose one of them as a key column to merge all datasets. I want to choose "GeoID" as a key column and drop "GeoType" and "GeoRank" columns. Because these column doesn't have any valuable data.

In [469]:
# List the columns that want to drop
columns_to_drop=["GeoType","GeoRank"]


In [476]:
# Create a new DataFrame with only the necessary columns

asthma_adults_df1= asthma_adults_df.drop(columns_to_drop, axis=1)

asthma_age_4_df1=asthma_age_4_df.drop(columns_to_drop,axis=1)

asthma_age5_17_df1=asthma_age5_17_df.drop(columns_to_drop, axis=1)

pm2_5_df1= pm2_5_df.drop(columns_to_drop,axis=1)

no2_df1= no2_df.drop(columns_to_drop,axis=1)

In [477]:
print(asthma_age_4_df1.columns)


Index(['TimePeriod', 'GeoID', 'Geography', 'Estimated annual rate per 10,000',
       'Number'],
      dtype='object')


Columns "GeoType" and "GeoRank" have removed from dataframe.

In [478]:
#rename the column 'TimePeriod' to 'Year'
asthma_adults_df1= asthma_adults_df1.rename(columns={"TimePeriod":"Year"})

asthma_age_4_df1=asthma_age_4_df1.rename(columns={"TimePeriod":"Year"})

asthma_age5_17_df1=asthma_age5_17_df1.rename(columns={"TimePeriod":"Year"})

pm2_5_df1= pm2_5_df1.rename(columns={"TimePeriod":"Year"})

no2_df1= no2_df1.rename(columns={"TimePeriod":"Year"})

In [479]:
no2_df1.head()

Year  GeoID      Geography  10th percentile ppb  90th percentile ppb  \
137  2022      3      Manhattan                 16.0                 23.3   
139  2022      5  Staten Island                  8.4                 13.9   
140  2022      1  New York City                 10.8                 18.1   
560  2021      3      Manhattan                 17.0                 23.7   
562  2021      5  Staten Island                 10.1                 15.8   

     Mean ppb  
137      19.1  
139      11.2  
140      14.7  
560      20.0  
562      12.8

In [480]:
asthma_age_4_df1.head()

Year  GeoID      Geography  Estimated annual rate per 10,000  Number
103  2020      3      Manhattan                              58.6     454
105  2020      5  Staten Island                              33.9      94
106  2020      1  New York City                              62.4    3357
210  2019      3      Manhattan                             193.0    1551
212  2019      5  Staten Island                              91.8     264

In [492]:
asthma_age_4_df1.isna().sum()

Year_ag_4                           0
GeoID                               0
Geography_ag_4                      0
Estimated annual rate per 10,000    0
Number                              0
dtype: int64

**4.Joining Data**

In [496]:
# rename the common columns except 'GeoID' to merge dataframes
asthma_adults_df1= asthma_adults_df1.rename(columns={'Geography':'Geography_ad',"Year":"Year_ad"})

asthma_age_4_df1= asthma_age_4_df1.rename(columns={'Geography':'Geography_ag_4',"Year":"Year_ag_4"})

asthma_age5_17_df1= asthma_age5_17_df1.rename(columns={'Geography':'Geography_ag5_17',"Year":"Year5_17"})

no2_df1=no2_df1.rename(columns={'Geography':'Geography_no2',"Year":"Year_no"})

pm2_5_df1=pm2_5_df1.rename(columns={'Geography':'Geography_pm', "Year":"Year_pm"})

In [497]:
merge1=pd.merge(asthma_adults_df1,asthma_age_4_df1, on="GeoID", how="inner")


In [498]:
merge2= pd.merge(merge1,asthma_age5_17_df1, on="GeoID", how="inner")

In [499]:
merge3= pd.merge(merge2,pm2_5_df1, on="GeoID", how="inner")

In [500]:
merge4=pd.merge(merge3,no2_df1, on="GeoID", how="inner")

In [501]:
merge4.head()

Year_ad  GeoID Geography_ad  Age-adjusted rate per 10,000  \
0     2020      3    Manhattan                          44.7   
1     2020      3    Manhattan                          44.7   
2     2020      3    Manhattan                          44.7   
3     2020      3    Manhattan                          44.7   
4     2020      3    Manhattan                          44.7   

   Estimated annual rate per 10,000_x  Number_x  Year_ag_4 Geography_ag_4  \
0                                43.5    6290.0       2020      Manhattan   
1                                43.5    6290.0       2020      Manhattan   
2                                43.5    6290.0       2020      Manhattan   
3                                43.5    6290.0       2020      Manhattan   
4                                43.5    6290.0       2020      Manhattan   

   Estimated annual rate per 10,000_y  Number_y  ...  Year_pm Geography_pm  \
0                                58.6       454  ...     2022    Manhattan   
1                                58.6       454  ...     2022    Manhattan   
2                                58.6       454  ...     2022    Manhattan   
3                                58.6       454  ...     2022    Manhattan   
4                                58.6       454  ...     2022    Manhattan   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  Year_no  \
0                     6.1                     8.8          7.0     2022   
1                     6.1                     8.8          7.0     2021   
2                     6.1                     8.8          7.0     2020   
3                     6.1                     8.8          7.0     2019   
4                     6.1                     8.8          7.0     2018   

  Geography_no2  10th percentile ppb  90th percentile ppb  Mean ppb  
0     Manhattan                 16.0                 23.3      19.1  
1     Manhattan                 17.0                 23.7      20.0  
2     Manhattan                 17.7                 24.4      20.5  
3     Manhattan                 17.7                 26.6      21.7  
4     Manhattan                 19.4                 28.8      23.4  

[5 rows x 25 columns]

In [502]:
merge4.columns

Index(['Year_ad', 'GeoID', 'Geography_ad', 'Age-adjusted rate per 10,000',
       'Estimated annual rate per 10,000_x', 'Number_x', 'Year_ag_4',
       'Geography_ag_4', 'Estimated annual rate per 10,000_y', 'Number_y',
       'Year5_17', 'Geography_ag5_17', 'Estimated annual rate per 10,000',
       'Number', 'small_number_estimate', 'Year_pm', 'Geography_pm',
       '10th percentile mcg/m3', '90th percentile mcg/m3', 'Mean mcg/m3',
       'Year_no', 'Geography_no2', '10th percentile ppb',
       '90th percentile ppb', 'Mean ppb'],
      dtype='object')

Results show some duplicated columns that have to be remove.

In [503]:
# Duplicated columns to drop
columns_to_remove=["Year_ag_4","Geography_ag_4","Year_pm","Geography_pm","Year_no","Geography_no2",'Geography_ag5_17',"Year5_17"]

In [504]:
merge4= merge4.drop(columns_to_remove, axis=1)

In [505]:
merge4.head()

Year_ad  GeoID Geography_ad  Age-adjusted rate per 10,000  \
0     2020      3    Manhattan                          44.7   
1     2020      3    Manhattan                          44.7   
2     2020      3    Manhattan                          44.7   
3     2020      3    Manhattan                          44.7   
4     2020      3    Manhattan                          44.7   

   Estimated annual rate per 10,000_x  Number_x  \
0                                43.5    6290.0   
1                                43.5    6290.0   
2                                43.5    6290.0   
3                                43.5    6290.0   
4                                43.5    6290.0   

   Estimated annual rate per 10,000_y  Number_y  \
0                                58.6       454   
1                                58.6       454   
2                                58.6       454   
3                                58.6       454   
4                                58.6       454   

   Estimated annual rate per 10,000  Number  small_number_estimate  \
0                              65.9    1088                  False   
1                              65.9    1088                  False   
2                              65.9    1088                  False   
3                              65.9    1088                  False   
4                              65.9    1088                  False   

   10th percentile mcg/m3  90th percentile mcg/m3  Mean mcg/m3  \
0                     6.1                     8.8          7.0   
1                     6.1                     8.8          7.0   
2                     6.1                     8.8          7.0   
3                     6.1                     8.8          7.0   
4                     6.1                     8.8          7.0   

   10th percentile ppb  90th percentile ppb  Mean ppb  
0                 16.0                 23.3      19.1  
1                 17.0                 23.7      20.0  
2                 17.7                 24.4      20.5  
3                 17.7                 26.6      21.7  
4                 19.4                 28.8      23.4

In [506]:
# Rename some of column's name
merge4= merge4.rename(columns={"Year_ad":"Year", "Geography_ad":"Geography","Age-adjusted rate per 10,000":"ad_per 10,000",
                              "Estimated annual rate per 10,000_x":"annual_ad per 10,000 ","Number_x":"Number_ad",
                               "Number_y":"Number_ag4", "Estimated annual rate per 10,000_y":"annual_ag4 per 10,000","Estimated annual rate per 10,000":"annual5_17 per 10,000",
                              "Number":"Number5_17","Mean mcg/m3":"pm/Mean mcg/m3","Mean ppb":"no2/Mean ppb"})

In [507]:
merge4.head()

Year  GeoID  Geography  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0  2020      3  Manhattan           44.7                   43.5     6290.0   
1  2020      3  Manhattan           44.7                   43.5     6290.0   
2  2020      3  Manhattan           44.7                   43.5     6290.0   
3  2020      3  Manhattan           44.7                   43.5     6290.0   
4  2020      3  Manhattan           44.7                   43.5     6290.0   

   annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                   58.6         454                   65.9        1088   
1                   58.6         454                   65.9        1088   
2                   58.6         454                   65.9        1088   
3                   58.6         454                   65.9        1088   
4                   58.6         454                   65.9        1088   

   small_number_estimate  10th percentile mcg/m3  90th percentile mcg/m3  \
0                  False                     6.1                     8.8   
1                  False                     6.1                     8.8   
2                  False                     6.1                     8.8   
3                  False                     6.1                     8.8   
4                  False                     6.1                     8.8   

   pm/Mean mcg/m3  10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0             7.0                 16.0                 23.3          19.1  
1             7.0                 17.0                 23.7          20.0  
2             7.0                 17.7                 24.4          20.5  
3             7.0                 17.7                 26.6          21.7  
4             7.0                 19.4                 28.8          23.4

In [508]:
merge4.shape

(2037616, 17)

In [509]:
# Using .ifo() to check Dtype of our data
merge4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2037616 entries, 0 to 2037615
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Year                    int64  
 1   GeoID                   int64  
 2   Geography               object 
 3   ad_per 10,000           float64
 4   annual_ad per 10,000    float64
 5   Number_ad               float64
 6   annual_ag4 per 10,000   float64
 7   Number_ag4              int64  
 8   annual5_17 per 10,000   float64
 9   Number5_17              int64  
 10  small_number_estimate   bool   
 11  10th percentile mcg/m3  float64
 12  90th percentile mcg/m3  float64
 13  pm/Mean mcg/m3          float64
 14  10th percentile ppb     float64
 15  90th percentile ppb     float64
 16  no2/Mean ppb            float64
dtypes: bool(1), float64(11), int64(4), object(1)
memory usage: 266.2+ MB


Based on the .info() columns 'annual_ag4 per 10,000' and ' Number_ag4' changed their datatype into object again. So we need to chnage to numeric again.

In [510]:
merge4['annual_ag4 per 10,000']=pd.to_numeric(merge4['annual_ag4 per 10,000'],errors='coerce')

In [511]:
merge4['Number_ag4']=pd.to_numeric(merge4['Number_ag4'],errors='coerce')

In [512]:
merge4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2037616 entries, 0 to 2037615
Data columns (total 17 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Year                    int64  
 1   GeoID                   int64  
 2   Geography               object 
 3   ad_per 10,000           float64
 4   annual_ad per 10,000    float64
 5   Number_ad               float64
 6   annual_ag4 per 10,000   float64
 7   Number_ag4              int64  
 8   annual5_17 per 10,000   float64
 9   Number5_17              int64  
 10  small_number_estimate   bool   
 11  10th percentile mcg/m3  float64
 12  90th percentile mcg/m3  float64
 13  pm/Mean mcg/m3          float64
 14  10th percentile ppb     float64
 15  90th percentile ppb     float64
 16  no2/Mean ppb            float64
dtypes: bool(1), float64(11), int64(4), object(1)
memory usage: 266.2+ MB


In [514]:
# calculates the sum of NA (missing) values.
merge4.isna().sum()

Year                      0
GeoID                     0
Geography                 0
ad_per 10,000             0
annual_ad per 10,000      0
Number_ad                 0
annual_ag4 per 10,000     0
Number_ag4                0
annual5_17 per 10,000     0
Number5_17                0
small_number_estimate     0
10th percentile mcg/m3    0
90th percentile mcg/m3    0
pm/Mean mcg/m3            0
10th percentile ppb       0
90th percentile ppb       0
no2/Mean ppb              0
dtype: int64

There is no any NaN values in the dataframe.

In [515]:
# calculates the sum of duplicate values
merge4.duplicated().sum()

57344

There are 57344 duplicate data in merged dataset.

**5.Numerical Features**

In [516]:
# Filter columns with number data to claculate some statistic such as: mean,max,min,...
merge4.select_dtypes(include=["number"])

Year  GeoID  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0        2020      3           44.7                   43.5     6290.0   
1        2020      3           44.7                   43.5     6290.0   
2        2020      3           44.7                   43.5     6290.0   
3        2020      3           44.7                   43.5     6290.0   
4        2020      3           44.7                   43.5     6290.0   
...       ...    ...            ...                    ...        ...   
2037611  2005      1           99.5                  100.3    62425.0   
2037612  2005      1           99.5                  100.3    62425.0   
2037613  2005      1           99.5                  100.3    62425.0   
2037614  2005      1           99.5                  100.3    62425.0   
2037615  2005      1           99.5                  100.3    62425.0   

         annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                         58.6         454                   65.9        1088   
1                         58.6         454                   65.9        1088   
2                         58.6         454                   65.9        1088   
3                         58.6         454                   65.9        1088   
4                         58.6         454                   65.9        1088   
...                        ...         ...                    ...         ...   
2037611                  377.2       19927                  199.2       25834   
2037612                  377.2       19927                  199.2       25834   
2037613                  377.2       19927                  199.2       25834   
2037614                  377.2       19927                  199.2       25834   
2037615                  377.2       19927                  199.2       25834   

         10th percentile mcg/m3  90th percentile mcg/m3  pm/Mean mcg/m3  \
0                           6.1                     8.8             7.0   
1                           6.1                     8.8             7.0   
2                           6.1                     8.8             7.0   
3                           6.1                     8.8             7.0   
4                           6.1                     8.8             7.0   
...                         ...                     ...             ...   
2037611                     9.3                    11.9            10.4   
2037612                     9.3                    11.9            10.4   
2037613                     9.3                    11.9            10.4   
2037614                     9.3                    11.9            10.4   
2037615                     9.3                    11.9            10.4   

         10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0                       16.0                 23.3          19.1  
1                       17.0                 23.7          20.0  
2                       17.7                 24.4          20.5  
3                       17.7                 26.6          21.7  
4                       19.4                 28.8          23.4  
...                      ...                  ...           ...  
2037611                 12.6                 23.6          18.8  
2037612                 13.3                 24.3          19.6  
2037613                 14.7                 26.7          21.6  
2037614                 14.1                 26.3          21.1  
2037615                 15.4                 28.6          22.7  

[2037616 rows x 15 columns]

In [519]:
#Creat descriptive statistics of number of asthma for adults 
merge4["Number_ad"].describe()

count    2.037616e+06
mean     2.913874e+04
std      2.871973e+04
min      1.287000e+03
25%      2.801000e+03
50%      1.429400e+04
75%      6.242500e+04
max      7.779200e+04
Name: Number_ad, dtype: float64

In [520]:
#Creat descriptive statistics of number of asthma for children up to 4 years.
merge4["Number_ag4"].describe()

count    2.037616e+06
mean     7.660372e+03
std      8.247117e+03
min      9.400000e+01
25%      5.280000e+02
50%      2.982000e+03
75%      1.722800e+04
max      2.149200e+04
Name: Number_ag4, dtype: float64

In [521]:
#Creat descriptive statistics of number of asthma for children aged 5 to 17 years.
merge4["Number5_17"].describe()

count    2.037616e+06
mean     1.128678e+04
std      1.121299e+04
min      2.180000e+02
25%      8.630000e+02
50%      4.594000e+03
75%      2.510000e+04
max      2.837000e+04
Name: Number5_17, dtype: float64

In [522]:
#Creat descriptive statistics of PM2.5.
merge4["pm/Mean mcg/m3"].describe()

count    2.037616e+06
mean     8.297373e+00
std      1.793996e+00
min      5.200000e+00
25%      6.800000e+00
50%      8.350000e+00
75%      9.500000e+00
max      1.260000e+01
Name: pm/Mean mcg/m3, dtype: float64

In [523]:
#Creat descriptive statistics of no2.
merge4["no2/Mean ppb"].describe()

count    2.037616e+06
mean     1.906507e+01
std      5.516172e+00
min      1.120000e+01
25%      1.470000e+01
50%      1.810000e+01
75%      2.170000e+01
max      3.390000e+01
Name: no2/Mean ppb, dtype: float64

 **6.Categorical Features**

In [524]:
merge4.select_dtypes(include=['object'])


Geography
0            Manhattan
1            Manhattan
2            Manhattan
3            Manhattan
4            Manhattan
...                ...
2037611  New York City
2037612  New York City
2037613  New York City
2037614  New York City
2037615  New York City

[2037616 rows x 1 columns]

In [525]:
merge4.shape

(2037616, 17)

In [526]:
merge4.head()

Year  GeoID  Geography  ad_per 10,000  annual_ad per 10,000   Number_ad  \
0  2020      3  Manhattan           44.7                   43.5     6290.0   
1  2020      3  Manhattan           44.7                   43.5     6290.0   
2  2020      3  Manhattan           44.7                   43.5     6290.0   
3  2020      3  Manhattan           44.7                   43.5     6290.0   
4  2020      3  Manhattan           44.7                   43.5     6290.0   

   annual_ag4 per 10,000  Number_ag4  annual5_17 per 10,000  Number5_17  \
0                   58.6         454                   65.9        1088   
1                   58.6         454                   65.9        1088   
2                   58.6         454                   65.9        1088   
3                   58.6         454                   65.9        1088   
4                   58.6         454                   65.9        1088   

   small_number_estimate  10th percentile mcg/m3  90th percentile mcg/m3  \
0                  False                     6.1                     8.8   
1                  False                     6.1                     8.8   
2                  False                     6.1                     8.8   
3                  False                     6.1                     8.8   
4                  False                     6.1                     8.8   

   pm/Mean mcg/m3  10th percentile ppb  90th percentile ppb  no2/Mean ppb  
0             7.0                 16.0                 23.3          19.1  
1             7.0                 17.0                 23.7          20.0  
2             7.0                 17.7                 24.4          20.5  
3             7.0                 17.7                 26.6          21.7  
4             7.0                 19.4                 28.8          23.4

**7.Save data**

In [527]:
# save the data to a new csv file
# Path to the directory where you want to save the file
data_path = '../data'
# Create the directory if it does not exist
os.makedirs(data_path, exist_ok=True)

#save the dataframe to a CSV file in the new directory
merge4.to_csv(os.path.join(data_path, 'merge4_cleaned.csv'), index=False)
